Return an error when "gid_from_name" is set but group does not exist

meaksh opened the pull request:
### What does this PR do?
This PR reports an error running `user.present` state in cases where `gid_from_name` is set to True but there is no group on the system with that name.

Also this prevents from wrongly take an empty string as a possible value for a group. It should be handled as `None` on that case.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/45345

### Previous Behavior
Given this SLS file:
```yaml
test:
  user.present:
    - gid_from_name: True
```
Applying the state:
```
local:
----------
          ID: test
    Function: user.present
      Result: True
     Comment: New user test created
     Started: 12:50:46.306964
    Duration: 251.407 ms
     Changes:   
              ----------
              fullname:
              gid:
                  100
              groups:
                  - users
              home:
                  /home/test
              homephone:
              name:
                  test
              passwd:
                  x
              roomnumber:
              shell:
                  /bin/bash
              uid:
                  1000
              workphone:

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 251.407 ms
```
As you see, the user is created but the group `test` doesn't. It set the gid to `users` groups instead of a group called `test` (according to the SLS definition).

If I run it again the state I got this failure as Salt tries to handle an empty string as a gid:
```
          ID: test
    Function: user.present
      Result: False
     Comment: These values could not be changed: {'gid': ''}
     Started: 13:39:51.155928
    Duration: 18.243 ms
     Changes:
```

### New Behavior
With the new behavior, the state will fail initially as the `test` group is not available on the system:

```
local:
----------
          ID: test
    Function: user.present
      Result: False
     Comment: Default group with name "test" is not present
     Started: 12:54:16.301860
    Duration: 8.197 ms
     Changes:   

Summary for local
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:   8.197 ms
```

Here we follow the general behavior of `user.present`, which is not creating groups in case they don't exist but reporting an error telling the groups don't exist.

### Tests written?

Yes

### Commits signed with GPG?

Yes

meaksh committed:
Ensure empty string gid is set to None

meaksh committed:
Return error when gid_from_name and group does not exist.

Fixes #45345

meaksh committed:
Make pylint happy

rallytime commented:
Hi @meaksh - thank you for submitting this fix.

There is a related test failing here in the PY3 run: `integration.states.test_user.UserTest.test_user_present_gid_from_name_default`

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/5546/

meaksh committed:
Fix integration tests for 'user.present' state.

meaksh committed:
Update documentation for 'gid_from_name' parameter

isbm commented on salt/states/user.py:
```diff
@@ -88,6 +88,10 @@ def _changes(name,
 
+    # Set gid to None in cases an empty string is passed
+    if gid == '':
+        gid = None
```
Move this to `def present` right below the line 462 where `__salt__['file.group_to_gid']...` is called, since you already checking it below on lines 463-467.

isbm commented on salt/states/user.py:
```diff
@@ -455,6 +460,11 @@ def present(name,
 
     if gid_from_name:
         gid = __salt__['file.group_to_gid'](name)
+        if not gid:
```
This is a logical bug (it will never actually happen in real life, because you always have 'root:root'): you say here if `gid` is `int("0")` and so is `False` in this case, the code meant to bailout, which is incorrect. I would then do so:

```python
if git_from_name:
    gid = __salt__['file.group_to_gid'](name)
    if gid == '':
        gid = None
    # ^
    # |
    # +--- at this point you've finished with setting
    #      a proper type of "gid" once and for all here.
    #
    # Now bailout if things are wrong with the default groups:
    if gid is None:
        ret['comment'] = 'be be be....'  # Please keep this in one line
        ...
        return ret
```

In this case function `_changes` (called twice in this scope) will get your `gid=None` instead of `gid=''` and so the issue is fine. But you have no scattered checks all around the place with a semi-bug introduced.

isbm reviewed (changes requested):
@meaksh please look at these.

isbm commented:
@rallytime the Git says "changes approved" but actually no, not really. This is confusing...

rallytime commented:
@isbm Can you elaborate on what you mean?

meaksh committed:
Refactor to prevent logical bug when gid is 0

meaksh commented:
@isbm - I updated the PR to prevent the logical bug when the returned gid is 0, but also refactored a bit more:

We only call `__salt__['file.group_to_gid'](name)` when `gid_from_name == True', and this call always returns either an `int` when the group exists or empty string if not. We don't need to fix the `gid` in case of an empty string here, if there is such empty string we just return with an error. That's it.

isbm reviewed (approved):
@meaksh see? I always say: less code = less bugs. :wink: !

isbm commented:
@rallytime this is now good to go!
